### PR TITLE
[Spark] Support more than 2 keys in spark online target ingestion engine

### PR DESCRIPTION
**Pull request by alxtkr77**
[ML-3443](https://jira.iguazeng.com/browse/ML-3443)

### CODE REVIEW

File: tests/system/feature_store/test_spark_engine.py [hunk 1]
```diff
@@ -383,6 +383,63 @@ def test_ingest_to_redis(self):
...
+        )
+        source = ParquetSource("myparquet", path=self.get_remote_pq_source_path())
+        if target_kind == "Redis":
+            targets = [RedisNoSqlTarget()]
+            measurements.set_targets(targets, with_defaults=False)
+        fstore.ingest(
+            measurements,
+            source,
+            spark_context=self.spark_service,
+            run_config=fstore.RunConfig(False),
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/system/feature_store/test_spark_engine.py:410]: Actually its '(function=False)'. Weird, but I copied it from other tests...

File: tests/system/feature_store/test_spark_engine.py [hunk 2]
```diff
@@ -383,6 +383,63 @@ def test_ingest_to_redis(self):
...
+            overwrite=True,
+        )
+        # read the dataframe from the redis back
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/system/feature_store/test_spark_engine.py:413]: Right..

File: tests/system/feature_store/test_spark_engine.py [hunk 3]
```diff
@@ -383,6 +383,63 @@ def test_ingest_to_redis(self):
...
+        measurements = fstore.FeatureSet(
+            name,
+            entities=[fstore.Entity(key1), fstore.Entity(key2), fstore.Entity(key3)],
+            timestamp_key="timestamp",
+            engine="spark",
+        )
+        source = ParquetSource("myparquet", path=self.get_remote_pq_source_path())
+        if target_kind == "Redis":
+            targets = [RedisNoSqlTarget()]
+            measurements.set_targets(targets, with_defaults=False)
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/system/feature_store/test_spark_engine.py:405]: Ok

File: tests/system/feature_store/test_spark_engine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gtopper) [reply tests/system/feature_store/test_spark_engine.py:413]: Yeah, but that's wrong. The parameter `function` isn't even supposed to be of type boolean. I guess those other tests are wrong too :)

File: tests/system/feature_store/test_spark_engine.py [hunk 4]
```diff
@@ -383,6 +383,66 @@ def test_ingest_to_redis(self):
...
+        measurements.set_targets(targets, with_defaults=False)
+
+        fstore.ingest(
+            measurements,
+            source,
+            spark_context=self.spark_service,
+            run_config=fstore.RunConfig(False),
+            overwrite=True,
+        )
+        # read the dataframe
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py:416]: I don't think this is reading any dataframe.
```suggestion
```

File: tests/system/feature_store/test_spark_engine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gtopper) [reply tests/system/feature_store/test_spark_engine.py:413]: There you go https://github.com/mlrun/mlrun/pull/3363

**Review by gtopper — APPROVED** [review]

File: mlrun/datastore/targets.py [hunk 5]
```diff
@@ -1056,13 +1057,13 @@ def get_spark_options(self, key_column=None, timestamp_key=None, overwrite=True)
...
-                raise mlrun.errors.MLRunInvalidArgumentError(
-                    f"Spark supports maximun of 2 keys and {key_column} are provided"
-                )
             spark_options["key"] = key_column[0]
-            if len(key_column) > 1:
+            if len(key_column) > 2:
+                spark_options["sorting-key"] = "_spark_object_sorting_key"
+            if len(key_column) == 2:
                 spark_options["sorting-key"] = key_column[1]
+            if len(key_column) < 2:
```
Comment (assaf758) [inline mlrun/datastore/targets.py:1065]: seems redundant - spark_options["key"] is set above

File: mlrun/datastore/targets.py [hunk 6]
```diff
@@ -1076,12 +1077,21 @@ def as_df(self, columns=None, df_module=None, **kwargs):
         raise NotImplementedError()
 
     def prepare_spark_df(self, df, key_columns):
```
Comment (assaf758) [inline mlrun/datastore/targets.py:1079]: please move to class NoSqlTarget (along with get_spark_options)

**Review by assaf758 — APPROVED** [review]

**Pull request merged**